week-four/app/src/main/java/pro/marvinhosea/movielist/ui/movies/MovieListFragment.kt

```diff
@@ -0,0 +1,164 @@
+
+    private val remoteApi by lazy {
+        RemoteApi(apiService)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setHasOptionsMenu(true)
+    }
+
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
+        super.onActivityCreated(savedInstanceState)
+        movieRecyclerView.adapter = adapter
+        if (!networkStatusChecker?.hasInternetConnection()!!) {
+            requireContext().toast("No Internet Connection")
+            lifecycleScope.launch {
+                adapter.setData(movieViewModel.getAllMovies())
+            }
+            return
+        }
+        getMovies("Top Rated Movies")
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
+        inflater.inflate(R.menu.menu_main, menu)
+        super.onCreateOptionsMenu(menu, inflater)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        when (item.itemId) {
+            R.id.watch_list -> myWatchList()
+            R.id.top_rate -> getMovies("Top Rated Movies")
+            R.id.popular_movies -> getMovies("Popular movies")
+            R.id.now_playing -> getMovies("Now Playing Movies")
+            R.id.upcoming_movies -> getMovies("Upcoming Movies")
+            R.id.logout_menu -> logOut()
+        }
+        return super.onOptionsItemSelected(item)
+    }
+
+    private fun myWatchList() {
+        lifecycleScope.launch {
+            (activity as AppCompatActivity).supportActionBar?.title = "My Watchlist"
+            adapter.setData(movieViewModel.myWatchListMovies())
+        }
+    }
+
+    private fun logOut() {
+        activity?.applicationContext?.let { userSharedRepository.init(it) }
+        userSharedRepository.logoutUser()
+
+        val intent = Intent(activity, UserLoginActivity::class.java)
+        startActivity(intent)
+        activity?.finish()
+    }
+
+    private fun getMovies(category: String) {
+        lifecycleScope.launch {
+            val results = when (category) {
+                "Upcoming Movies" -> remoteApi.getUpcomingMovies()
+                "Top Rated Movies" -> remoteApi.getTopRatedMovies()
+                "Now Playing Movies" -> remoteApi.getNowPlayingMovies()
+                else -> {
+                    remoteApi.getPopularMovies()
+                }
+            }
+
+            if (results is Success) {
+                val movies = formatResponseMovies(results.data)
+
+                adapter.setData(movies)
+                (activity as AppCompatActivity).supportActionBar?.title = category
+                movieViewModel.saveMovies(movies)
+            } else {
+                Log.d("Error1", "testing ${results.toString()}")
+                Toast.makeText(activity, "Problem loading movies", Toast.LENGTH_LONG).show()
+            }
+        }
+    }
+
+    private fun formatResponseMovies(moviesResponse: List<Result>): List<Movie> {
```

Comment:
Same here business logic, should be in the ViewModel if possible.

Also, so many different parts of the function, for one, I'm not sure why you keep checking the username if it's empty or not.

If the user reached this Fragment and it's showing, then there is no way to log out, without moving away from this fragment. It's redundant to keep checking that, if you're already checking it in two places in the MainActivity.